code/headers/hardware_usart.hpp

```diff
@@ -2,76 +2,184 @@
+
+            hardware_usart->US_MR =
+                    UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
+            hardware_usart->US_IDR = 0xFFFFFFFF;
+
+            enable();
+        }
+
         /// @brief char output operator
         ///
         /// Although calling send_byte should do the exact same thing.
         /// In practice useing this fuction is more stable
         /// Especially when the values are repeated.
-        hardware_usart_c &operator<<(uint8_t byte);
+        hardware_usart_c &operator<<(uint8_t byte) {
+            send_byte(byte);
+            return *this;
+        }
 
         /// @brief string output operator
         ///
         /// used for char arrays to quickly send more than one byte
         /// In practice useing this fuction is more stable than repeated use of
         /// send_byte() Especially when the values are repeated.
-        hardware_usart_c &operator<<(const char *c) ;
+        hardware_usart_c &operator<<(const char *c) {
+            for (const char *p = c; *p != '\0'; p++) {
+                send_byte(*p);
+            }
+            return *this;
+        }
 
         /// @brief enables the internal USART controller
-        void enable() override;
+        void enable() override {
+            hardware_usart->US_CR = UART_CR_RXEN | UART_CR_TXEN;
+        }
 
         /// @brief disables the internal USART controller
-        void disable() override;
+        void disable() override {
+            hardware_usart->US_CR =
+                    UART_CR_RSTRX | UART_CR_RSTTX | UART_CR_RXDIS | UART_CR_TXDIS;
+        }
 
         /// @brief send single byte
         /// @param c: byte to send
         /// @return true if byte send, false if not send
-        bool send(const uint8_t c) override;
+        bool send(const uint8_t c) override {
+           send_byte(c);
+           return true;
+        }
 
         /// @brief sends a char via usart
         /// @param c: char to send
-        void putc(char c) override;
+        void putc(char c) override {
+            send_byte(c);
+        }
 
         /// @brief recieve byte bia usart
         /// @return received byte
-        uint8_t receive() override;
+        uint8_t receive() override {
+            if (!input_buffer.size()) {
+                return 0;
+            }
+
+            return input_buffer.copy_and_pop();
+        }
 
         /// @brief check if char is available
         /// @return true if char is available false if not
-        bool char_available() override;
+        bool char_available() override {
+            return (available() > 0);
+        }
 
         /// @brief receive char via usart
         /// @return char received
-        char getc() override;
+        char getc() override {
+            if (char_available()) {
+                return receive();
+            }
+            return 0;
+        }
 
         /// @brief returns available data in buffer
```

Comment:
This doesn't seem right, it doesn't actually return the available data, just the bytecount.

Reply:
on it